infrastructure/network/addressmanager/addressmanager.go

```diff
@@ -5,1264 +5,254 @@
-	newAddressBucketArray := am.newAddressBucketArray(knownAddress.subnetworkID)
-	newAddressBucketIndex := am.newAddressBucketIndex(netAddress, sourceAddress)
-	if newAddressBucketArray != nil {
-		if _, ok := newAddressBucketArray[newAddressBucketIndex][addressKey]; ok {
-			return
-		}
-	}
-
-	// Enforce max addresses.
-	if newAddressBucketArray != nil && len(newAddressBucketArray[newAddressBucketIndex]) > newBucketSize {
-		log.Tracef("new bucket is full, expiring old")
-		am.expireNew(knownAddress.subnetworkID, newAddressBucketIndex)
-	}
-
-	// Add to new bucket.
-	knownAddress.referenceCount++
-	am.updateAddrNew(newAddressBucketIndex, addressKey, knownAddress)
-
-	totalAddressCount := am.newAddressCount(knownAddress.subnetworkID) + am.triedAddressCount(knownAddress.subnetworkID)
-	log.Tracef("Added new address %s for a total of %d addresses", addressKey, totalAddressCount)
-
+// AcceptUnroutable specifies whether this network accepts unroutable
+// IP addresses, such as 10.0.0.0/8
+func (amc *Config) AcceptUnroutable() bool {
+	return amc.cfg.NetParams().AcceptUnroutable
 }
 
-func (am *AddressManager) updateAddrNew(bucket int, addressKey AddressKey, knownAddress *KnownAddress) {
-	if knownAddress.subnetworkID == nil {
-		am.fullNodeNewAddressBucketArray[bucket][addressKey] = knownAddress
-		return
-	}
-
-	if _, ok := am.subnetworkNewAddressBucketArrays[*knownAddress.subnetworkID]; !ok {
-		am.subnetworkNewAddressBucketArrays[*knownAddress.subnetworkID] = &newAddressBucketArray{}
-		for i := range am.subnetworkNewAddressBucketArrays[*knownAddress.subnetworkID] {
-			am.subnetworkNewAddressBucketArrays[*knownAddress.subnetworkID][i] = make(map[AddressKey]*KnownAddress)
-		}
-	}
-	am.subnetworkNewAddressBucketArrays[*knownAddress.subnetworkID][bucket][addressKey] = knownAddress
+// AddressManager provides a concurrency safe address manager for caching potential
+// peers on the Kaspa network.
+type AddressManager struct {
+	addresses map[AddressKey]*netAddressWrapper
+	mutex     sync.Mutex
+	cfg       Configer
+	random    AddressRandomizer
 }
 
-func (am *AddressManager) updateAddrTried(bucketIndex int, knownAddress *KnownAddress) {
-	if knownAddress.subnetworkID == nil {
-		am.fullNodeTriedAddressBucketArray[bucketIndex] = append(am.fullNodeTriedAddressBucketArray[bucketIndex], knownAddress)
-		return
-	}
-
-	if _, ok := am.subnetworkTriedAddresBucketArrays[*knownAddress.subnetworkID]; !ok {
-		am.subnetworkTriedAddresBucketArrays[*knownAddress.subnetworkID] = &triedAddressBucketArray{}
-		for i := range am.subnetworkTriedAddresBucketArrays[*knownAddress.subnetworkID] {
-			am.subnetworkTriedAddresBucketArrays[*knownAddress.subnetworkID][i] = nil
-		}
-	}
-	am.subnetworkTriedAddresBucketArrays[*knownAddress.subnetworkID][bucketIndex] = append(am.subnetworkTriedAddresBucketArrays[*knownAddress.subnetworkID][bucketIndex], knownAddress)
+type netAddressWrapper struct {
+	netAddress *appmessage.NetAddress
+	isBanned   bool
+	isLocal    bool
```

Comment:
I think it makes more sense to keep two lists: one for local, one for other addresses.
They are never used interchageably, and it will save us quite a lot of looping inside BestLocalAddress